File: schemas/src/synesis_schemas/main_server/pipeline/pipeline.py
from uuid import UUID
from datetime import datetime
from typing import Optional, Literal, List
from pydantic import BaseModel


class PipelineInDB(BaseModel):
    id: UUID
    user_id: UUID
    name: str
    created_at: datetime
    updated_at: datetime
    description: Optional[str] = None


class PipelineRunInDB(BaseModel):
    id: UUID
    pipeline_id: UUID
    status: Literal["pending", "running", "completed", "failed"]
    start_time: datetime
    end_time: datetime
    created_at: datetime
    updated_at: datetime


class FunctionInDB(BaseModel):
    id: UUID
    name: str
    implementation_script_path: str
    created_at: datetime
    updated_at: datetime
    description: str
    embedding: List[float]
    type: Literal["inference", "training", "computation"]
    setup_script_path: Optional[str] = None
    default_config: Optional[dict] = None


class FunctionInPipelineInDB(BaseModel):
    id: UUID
    pipeline_id: UUID
    function_id: UUID
    position: int
    created_at: datetime
    updated_at: datetime
    config: Optional[dict] = None

File: main_server/client/src/types/ontology/pipeline.ts
import { UUID } from "crypto";

export type PipelineRunStatus = "running" | "completed" | "failed";

// Base Models

export interface PipelineBase {
  id: UUID;
  userId: UUID;
  name: string;
  description: string;
  createdAt: string;
  updatedAt: string;
}

export interface PipelineImplementationBase {
  id: UUID;
  pythonFunctionName: string;
  docstring: string;
  description: string;
  argsSchema: Record<string, unknown>;
  defaultArgs: Record<string, unknown>;
  outputVariablesSchema: Record<string, unknown>;
  implementationScriptPath: string;
  createdAt: string;
  updatedAt: string;
}

export interface PipelineRunBase {
  id: UUID;
  args: Record<string, unknown>;
  pipelineId: UUID;
  outputVariables: Record<string, unknown>;
  name?: string | null;
  description?: string | null;
  status: PipelineRunStatus;
  startTime: string;
  endTime?: string | null;
  createdAt: string;
  updatedAt: string;
}

// Composite Models

export interface Pipeline extends PipelineBase {
  runs: PipelineRunBase[];
  implementation?: PipelineImplementationBase | null;
}

// Create Models

export interface PipelineImplementationCreate {
  pythonFunctionName: string;
  docstring: string;
  description: string;
  argsSchema: Record<string, unknown>;
  defaultArgs: Record<string, unknown>;
  outputVariablesSchema: Record<string, unknown>;
  implementationScriptPath: string;
  pipelineId: UUID;
}

export interface PipelineRunCreate {
  name: string;
  args: Record<string, unknown>;
  pipelineId: UUID;
  outputVariables?: Record<string, unknown>;
  description?: string | null;
  status?: PipelineRunStatus;
}

export interface PipelineCreate {
  name: string;
  description?: string | null;
  implementationCreate?: PipelineImplementationCreate | null;
  runsCreate?: PipelineRunCreate[] | null;
}

